NVorbis/DataPacket.cs

```diff
@@ -37,50 +18,99 @@ protected enum PacketFlags : byte
             /// </summary>
-            HasGranuleCount = 0x08,
-
+            User0 = 0x08,
             /// <summary>
             /// Flag for use by inheritors.
             /// </summary>
-            User1           = 0x10,
+            User1 = 0x10,
             /// <summary>
             /// Flag for use by inheritors.
             /// </summary>
-            User2           = 0x20,
+            User2 = 0x20,
             /// <summary>
             /// Flag for use by inheritors.
             /// </summary>
-            User3           = 0x40,
+            User3 = 0x40,
             /// <summary>
             /// Flag for use by inheritors.
             /// </summary>
-            User4           = 0x80,
+            User4 = 0x80,
+        }
+
+        ulong _bitBucket;
+        int _bitCount;
+        byte _overflowBits;
+        PacketFlags _packetFlags;
+        int _readBits;
+
+        /// <summary>
+        /// Gets the number of container overhead bits associated with this packet.
+        /// </summary>
+        public int ContainerOverheadBits { get; set; }
+
+        /// <summary>
+        /// Gets the granule position of the packet, if known.
+        /// </summary>
+        public long? GranulePosition { get; set; }
+
+        /// <summary>
+        /// Gets whether this packet occurs immediately following a loss of sync in the stream.
+        /// </summary>
+        public bool IsResync
+        {
+            get => GetFlag(PacketFlags.IsResync);
+            set => SetFlag(PacketFlags.IsResync, value);
+        }
+
+        /// <summary>
+        /// Gets whether this packet did not read its full data.
+        /// </summary>
+        public bool IsShort
+        {
+            get => GetFlag(PacketFlags.IsShort);
+            private set => SetFlag(PacketFlags.IsShort, value);
         }
 
         /// <summary>
-        /// Gets the value of the specified flag.
+        /// Gets whether the packet is the last packet of the stream.
         /// </summary>
-        protected bool GetFlag(PacketFlags flag)
+        public bool IsEndOfStream
         {
-            return (_packetFlags & flag) == flag;
+            get => GetFlag(PacketFlags.IsEndOfStream);
+            set => SetFlag(PacketFlags.IsEndOfStream, value);
         }
 
         /// <summary>
-        /// Sets the value of the specified flag.
+        /// Gets the number of bits read from the packet.
+        /// </summary>
+        public int BitsRead => _readBits;
+
+        /// <summary>
+        /// Gets the number of bits left in the packet.
+        /// </summary>
+        public int BitsRemaining => TotalBits - _readBits;
+
+        /// <summary>
+        /// Gets the total number of bits in the packet.
         /// </summary>
-        protected void SetFlag(PacketFlags flag, bool value)
+        abstract protected int TotalBits { get; }
+
+        bool GetFlag(PacketFlags flag) => (_packetFlags & flag) == flag;
```

Comment:
Isn't this just [HasFlag](https://docs.microsoft.com/en-us/dotnet/api/system.enum.hasflag?view=netframework-4.8)?

Reply:
Ha!  Good catch...  Probably not any faster, but is much cleaner.  Will be fixed shortly.